### PR TITLE
Add a new benchmark service

### LINKED ISSUE #92: Run Daily Benchmark Script Automagically
**Issue summary**

Tasks for running benchmark script automatically:
- [X] configure a droplet (detailed requirements listed below) that is automatically created daily and destroyed (destruction can either be triggered by the script or workflow, we may need to evaluate which would be best) after successful run of the script
- [X] move the benchmark script to `forest-iac` and deploy
- [X] configure droplet to export the CSV results to a location off-droplet
- [X] configure alerts if the script fails for any reason (including during setup and installation of pre-reqs)
- [x] test that everything works properly (creation/deletion of droplet, script runs or alerts on failure, CSV exports correctly)

**Other information and links**

We will need to automatically: create the droplet, install the pre-reqs as outlined in the `README.md`, run the benchmarks script (currently in the Forest repo [here](https://github.com/ChainSafe/forest/blob/44e150874293fb7f42d74681268678c8251d78a9/scripts/benchmark_db/bench.rb), needs to be moved to `forest-iac` as part of PR for this issue), export the CSV that is generated (possibly where our snapshots go?), and destroy the droplet upon successful run (or alert on failure without destroying the droplet to allow troubleshooting of the issue).

We need a machine with a dedicated CPU so the results are reproducible. A 320GB disk was sufficient to run the mainnet tests, but we may need more storage in the near future as features are added to the script. We have tested on both Ubuntu and Fedora droplets; as most of the existing IAC infra is Ubuntu, we can go with Ubuntu if we want to be consistent.

### PR DESCRIPTION
**Pull request by sudo-shashank**
**Summary of changes**
Changes introduced in this pull request:

Added a new benchmark service:
- Continuously runs benchmarks 
```
ruby bench.rb --chain calibnet --tempdir ./snapshots --daily
ruby bench.rb --chain mainnet --tempdir ./snapshots --daily
```
- Upload benchmark results to [`forest-benchmark`](https://cloud.digitalocean.com/spaces/forest-benchmarks?i=88c522) [weekly-results](https://cloud.digitalocean.com/spaces/forest-benchmarks?i=88c522&path=benchmark-results%2Fweekly-results%2F) and [all_results](https://forest-benchmarks.fra1.digitaloceanspaces.com/benchmark-results/all_results.csv) file.
- Sends regular status updates on slack `#forest-notifications` channel.

**Reference issue to close (if applicable)**

Closes #92 

**Other information and links**

### CODE REVIEW

File: benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.sh
Comment (LesnyRumcajs) [reply benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.sh:17]: @samuelarogbonlo It is fine to run them on the same host as long as this is not done sequentially and not in parallel, which seems to be the case here. The reasoning behind is that one Forest process should not steal resources from another Forest process.

File: benchmark_db_terraform/modules/benchmark_db/service/bench.rb
Comment (LesnyRumcajs) [reply benchmark_db_terraform/modules/benchmark_db/service/bench.rb:14]: It is a good point that we have a single source of truth for the benchmark script. On the other hand, this repository might be better suited for such code, so removing the original benchmarking code in the Forest repository is an alternative to what you're suggesting.

File: benchmark_db_terraform/main.tf
Comment (lemmih) [inline benchmark_db_terraform/main.tf:32]: I've created a new bucket for benchmark results: `forest-benchmarks`

File: benchmark_db_terraform/modules/benchmark_db/service/Gemfile.lock
Comment (lemmih) [inline benchmark_db_terraform/modules/benchmark_db/service/Gemfile.lock:20]: This is not the platform of the DO droplets.

File: benchmark_db_terraform/modules/benchmark_db/service/docker-compose.yml
Comment (lemmih) [inline benchmark_db_terraform/modules/benchmark_db/service/docker-compose.yml:25]: I don't understand these commands. Could you add some documentation here?

File: benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.rb
Comment (lemmih) [inline benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.rb:39]: Shouldn't these be part of the docker image?

File: benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.sh
Comment (lemmih) [inline benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.sh:21]: As we discussed in Slack, I would like it if the format was something like:
```
timestamp, forest-version, lotus-version, chain, metric, forest-value, lotus-value
```
We will add more metrics in the future.

**Review by lemmih — CHANGES_REQUESTED** [review]
Good start.

File: benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.sh
Comment (sudo-shashank) [reply benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.sh]: fixed

File: benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.rb
Comment (sudo-shashank) [reply benchmark_db_terraform/modules/benchmark_db/service/run_benchmark.rb]: yes moved

File: benchmark_db_terraform/modules/benchmark_db/service/Gemfile.lock
Comment (sudo-shashank) [reply benchmark_db_terraform/modules/benchmark_db/service/Gemfile.lock]: fixed

File: benchmark_db_terraform/main.tf
Comment (LesnyRumcajs) [inline benchmark_db_terraform/main.tf:31]: Let's not forget to change it.

File: benchmark_db_terraform/modules/benchmark_db/main.tf
Comment (LesnyRumcajs) [inline benchmark_db_terraform/modules/benchmark_db/main.tf:56]: Why do we need it?

File: benchmark_db_terraform/modules/benchmark_db/service/Dockerfile
Comment (LesnyRumcajs) [inline benchmark_db_terraform/modules/benchmark_db/service/Dockerfile:9]: Are you sure all of these are required to list? It'd be nice to group them and elaborate which package is a dependency of which feature.

File: benchmark_db_terraform/modules/benchmark_db/service/Dockerfile
Comment (LesnyRumcajs) [inline benchmark_db_terraform/modules/benchmark_db/service/Dockerfile:1]: It might be worth bumping it to `38`. Or using an Ubuntu LTS. While I do like the Fedora Project, its releases don't have long-term support.

File: benchmark_db_terraform/modules/benchmark_db/main.tf
Comment (sudo-shashank) [reply benchmark_db_terraform/modules/benchmark_db/main.tf:56]: removed unused envs

File: .github/workflows/deploy-benchmark-db.yml
Comment (lemmih) [inline .github/workflows/deploy-benchmark-db.yml:2]: We aren't just benchmarking the database.

File: benchmark_db_terraform/README.md
Comment (lemmih) [inline benchmark_db_terraform/README.md:4]: We're not testing database configurations. The benchmarking suite grew out of a tool that benchmarked database configurations, but we're benchmarking something else with this service.

File: benchmark_db_terraform/modules/benchmark_db/service/docker-compose.yml
Comment (lemmih) [inline benchmark_db_terraform/modules/benchmark_db/service/docker-compose.yml:22]: What does `wait` do here?

**Review by lemmih — CHANGES_REQUESTED** [review]
I need help understanding the logic. I will review it again once the PR description is updated.

File: benchmark_db_terraform/modules/benchmark_db/service/docker-compose.yml
Comment (sudo-shashank) [reply benchmark_db_terraform/modules/benchmark_db/service/docker-compose.yml]: fixed

File: .github/workflows/deploy-benchmark-db.yml
Comment (sudo-shashank) [reply .github/workflows/deploy-benchmark-db.yml]: changed to `Benchmark Service`

**Review by lemmih — CHANGES_REQUESTED** [review]
- Use a droplet image that already has docker installed (unless you have a reason to stay on fedora).
- Don't install docker inside of the docker container.
- You probably don't need `cmake` either.
- Handle secrets using: https://docs.docker.com/compose/use-secrets/

File: benchmark_terraform/modules/benchmark/service/docker-compose.yml
Comment (LesnyRumcajs) [inline benchmark_terraform/modules/benchmark/service/docker-compose.yml:1]: Why do we need a compose file for a single container?

File: benchmark_terraform/modules/benchmark/service/init.sh
Comment (LesnyRumcajs) [inline benchmark_terraform/modules/benchmark/service/init.sh:9]: `compose` is already available as a subcommand of `docker`. See https://github.com/ChainSafe/forest-iac/issues/124

File: benchmark_terraform/modules/benchmark/service/bench.rb
Comment (LesnyRumcajs) [inline benchmark_terraform/modules/benchmark/service/bench.rb:1]: Let's not forget to remove the [corresponding code](https://github.com/ChainSafe/forest/tree/main/scripts/benchmark_db) from Forest.

File: benchmark_terraform/modules/benchmark/service/.rubocop.yml
Comment (LesnyRumcajs) [inline benchmark_terraform/modules/benchmark/service/.rubocop.yml:1]: Let's introduce it globally for the entire repository.

File: benchmark_terraform/modules/benchmark/service/run_benchmark.sh
Comment (LesnyRumcajs) [inline benchmark_terraform/modules/benchmark/service/run_benchmark.sh:24]: what is the point of `||` at the end?
```suggestion
s3cmd get s3://"$BENCHMARK_BUCKET"/benchmark-results/weekly-results/weekly_result_"$week_number".csv /tmp/weekly_result_"$week_number".csv --force
```

**Review by LesnyRumcajs — COMMENTED** [review]
It'd be great to have instructions on running this manually on your own machine.

Like, imagine a case where there is a bug in the Ruby scripts (impossible!) and someone gets volunteered to fix them. I don't see clear instructions on how to do it here. Perhaps we could snatch the README (if it still applies) from Forest before removing it from there.

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
Please enable Rubocop for the introduced Ruby scripts. https://github.com/ChainSafe/forest-iac/blob/141793a8e188ac384dcdddf21a3f7f877228f949/.github/workflows/scripts-lint.yml#L31-L32

Have fun. :)

File: .github/workflows/deploy-benchmark.yml
Comment (LesnyRumcajs) [inline .github/workflows/deploy-benchmark.yml:4]: Let's also include paths so it doesn't get triggered when someone changes, e.g., the snapshot service.

Adding concurrency would also be helpful. Example here https://github.com/ChainSafe/forest-iac/blob/82884863aac0b04e4c9c55a4024564bdad0984e7/.github/workflows/deploy-forest-calibnet.yml#L4-L17

File: .github/workflows/scripts-lint.yml
Comment (LesnyRumcajs) [inline .github/workflows/scripts-lint.yml:32]: why did you remove it from `scripts`?

File: terraform/modules/benchmark/service/.rubocop.yml
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/.rubocop.yml:1]: Shouldn't we have a shared `.rubocop.yml` in the repo root?

File: terraform/modules/benchmark/service/Gemfile
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/Gemfile:1]: We either need to use Gemfile to install gems or manually install them via `gem install ..`. Is there a point in having those dependencies in both places? The gemfile is the preferred way.

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
Partial review, didn't check the Ruby source files yet.

File: .github/workflows/deploy-benchmark.yml
Comment (sudo-shashank) [reply .github/workflows/deploy-benchmark.yml:4]: done

File: .github/workflows/scripts-lint.yml
Comment (sudo-shashank) [reply .github/workflows/scripts-lint.yml:32]: fixed

File: terraform/modules/benchmark/service/.rubocop.yml
Comment (sudo-shashank) [reply terraform/modules/benchmark/service/.rubocop.yml:1]: yes that will be in another PR[ Enable rubocop for entire forest-iac repo](https://github.com/ChainSafe/forest-iac/pull/229 )

File: .github/workflows/deploy-benchmark.yml
Comment (LesnyRumcajs) [reply .github/workflows/deploy-benchmark.yml:4]: What about concurrency?

File: terraform/benchmark/README.md
Comment (LesnyRumcajs) [inline terraform/benchmark/README.md:5]: Which versions of Forest against which versions of Lotus? Latest releases? This should be specified in the README. Are the releases updated automatically?

File: terraform/modules/benchmark/service/Dockerfile
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/Dockerfile:12]: I'm a bit puzzled. Why do we need all of those dependencies? Could we just use Docker? We want to use the latest releases of both Forest and Lotus, which are conveniently dockerized. This would reduce the amount of manual work in the benchmark itself.

That said, I'm okay with doing this in a separate issue. If so, let's create it.

File: terraform/modules/benchmark/service/bench.rb
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/bench.rb:160]: It's wasteful to include the unit in every row of the result. This should be a part of the header.

File: terraform/modules/benchmark/service/run_benchmark.sh
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/run_benchmark.sh:30]: Is overwriting benchmarks from January 2023 with results from January 2024 expected and by design?

File: terraform/modules/benchmark/service/run_benchmark.sh
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/run_benchmark.sh:31]: the `/tmp/weekly_result_"$week_number".csv` is repeated many times. Let's use a variable.

File: terraform/modules/benchmark/service/run_benchmark.sh
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/run_benchmark.sh:33]: What is the plan to deal with breaking changes? Looking at the existing file at https://forest-benchmarks.fra1.digitaloceanspaces.com/benchmark-results/all_results.csv, there were some at one point:
```
2023-07-05 23:43:49,0.11.0,1.23.4,calibnet,validate_online,11.8,8.0
2023-07-06 01:48:32,0.11.0,1.23.4,mainnet,import,1443,2094
2023-07-06 01:48:32,0.11.0,1.23.4,mainnet,validate_online,0.209,0.175
2023-07-06 02:36:54,0.11.0,1.23.4,calibnet,import,41.1,96
2023-07-06 02:36:54,0.11.0,1.23.4,calibnet,validate_online,2.5,2.6
2023-07-06 04:43:39,0.11.0,1.23.4,mainnet,import,1480,2082
2023-07-06 04:43:39,0.11.0,1.23.4,mainnet,validate_online,0.275,0.184
2023-07-06 16:07:48,0.11.1,1.23.4,calibnet,import_time,44.8 sec,102 sec
2023-07-06 16:07:48,0.11.1,1.23.4,calibnet,validation_time,11.7 tipsets/sec,7.4 tipsets/sec
2023-07-06 18:15:00,0.11.1,1.23.4,mainnet,import_time,1415 sec,2054 sec
2023-07-06 18:15:00,0.11.1,1.23.4,mainnet,validation_time,0.159 tipsets/sec,0.092 tipsets/sec
2023-07-06 16:07:48,0.11.1,1.23.4,calibnet,import_time,44.8 sec,102 sec
2023-07-06 16:07:48,0.11.1,1.23.4,calibnet,validation_time,11.7 tipsets/sec,7.4 tipsets/sec
2023-07-06 18:15:00,0.11.1,1.23.4,mainnet,import_time,1415 sec,2054 sec
2023-07-06 18:15:00,0.11.1,1.23.4,mainnet,validation_time,0.159 tipsets/sec,0.092 tipsets/sec
2023-07-06 19:03:26,0.11.1,1.23.4,calibnet,import_time,43.6 sec,101 sec
2023-07-06 19:03:26,0.11.1,1.23.4,calibnet,validation_time,12.3 tipsets/sec,7.6 tipsets/sec
2023-07-06 21:07:41,0.11.1,1.23.4,mainnet,import_time,1400 sec,2110 sec
2023-07-06 21:07:41,0.11.1,1.23.4,mainnet,validation_time,0.267 tipsets/sec,0.192 tipsets/sec
2023-07-06 16:07:48,0.11.1,1.23.4,calibnet,import_time,44.8 sec,102 sec
```
Should the former ones be cleared? What is the plan for the future?

File: terraform/modules/benchmark/service/run_benchmark.sh
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/run_benchmark.sh:37]: Why do we need to `exit 1` in case of an error? Shouldn't `set -e` that is already in the file work fine?

File: terraform/modules/benchmark/service/run_benchmark.rb
Comment (LesnyRumcajs) [inline terraform/modules/benchmark/service/run_benchmark.rb:26]: What is sync and health and sync log exactly and how are they created / populated?

**Review by LesnyRumcajs — CHANGES_REQUESTED** [review]
I didn't look that much into Ruby scripts themselves, given they're mostly imported from Forest. All in all, it'd be good to have some more form in them and potentially unit tests (and a way to easily run them locally). We can think about it in another PR.

Another thing is compiling Rust and Lotus manually. This shouldn't be needed with Docker images and would simplify dependencies and initialization of the benchmark.

Also, I'd like to know the plans for CSV migrations, e.g., what do we do when a column is added/removed/modified in the CSV. We can no longer easily compare the results in such case anymore (well, adding a column should be fine).

File: terraform/benchmark/README.md
Comment (sudo-shashank) [reply terraform/benchmark/README.md:5]: corrected

File: terraform/modules/benchmark/service/Dockerfile
Comment (sudo-shashank) [reply terraform/modules/benchmark/service/Dockerfile:12]: I'll create a separate issue for this

File: terraform/modules/benchmark/service/bench.rb
Comment (sudo-shashank) [reply terraform/modules/benchmark/service/bench.rb:160]: `import_time` and `validation_time` comes under `Metric` header, each metric will have its own unit
we will have to restructure the output format 
I can take this out in a separate PR

File: terraform/modules/benchmark/service/run_benchmark.sh
Comment (sudo-shashank) [reply terraform/modules/benchmark/service/run_benchmark.sh:30]: added year to avoid this condition

File: terraform/modules/benchmark/service/run_benchmark.rb
Comment (sudo-shashank) [reply terraform/modules/benchmark/service/run_benchmark.rb:26]: fixed, now there is only one log file `benchmark_#{datetime}_run`

File: terraform/modules/benchmark/service/run_benchmark.sh
Comment (sudo-shashank) [reply terraform/modules/benchmark/service/run_benchmark.sh:31]: fixed

File: terraform/modules/benchmark/service/run_benchmark.sh
Comment (sudo-shashank) [reply terraform/modules/benchmark/service/run_benchmark.sh]: will take care in a separate PR

**Review by LesnyRumcajs — APPROVED** [review]
Please add the concurrency as mentioned in one of the comments. Otherwise, I'm okay with merging this, provided we address improvements in a separate PR soon.

**Comment by LesnyRumcajs** [comment]
> @lemmih @LesnyRumcajs don’t you guys think we should also step up new relic to monitor the droplet?

@samuelarogbonlo let's do it in a separate PR and not increase the load on NR until we are sure that the fixes you made land us well below the free tier limit.

**Comment by LesnyRumcajs** [comment]
By the way, shall we delete the [benchmark script from the Forest repository](https://github.com/ChainSafe/forest/tree/main/scripts/benchmark_db) now? @sudo-shashank

**Comment by sudo-shashank** [comment]
> By the way, shall we delete the [benchmark script from the Forest repository](https://github.com/ChainSafe/forest/tree/main/scripts/benchmark_db) now? @sudo-shashank

yes we can once this is merged

File: terraform/modules/benchmark/main.tf
Comment (sudo-shashank) [reply terraform/modules/benchmark/main.tf:75]: added to https://github.com/ChainSafe/forest-iac/issues/234

**Comment by sudo-shashank** [comment]
@lemmih can you approve this PR, its waiting for your approval to be merged

**Pull request merged**
